primitives/numeric/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+    struct NumericScaleHelper {
+        #[codec(compact)]
+        mantissa: u128,
+        #[codec(compact)]
+        scale: u32,
+    }
+
+    impl Encode for Numeric {
+        fn encode(&self) -> Vec<u8> {
+            NumericScaleHelper {
+                mantissa: self.mantissa(),
+                scale: self.scale(),
+            }
+            .encode()
+        }
+    }
+
+    impl Decode for Numeric {
+        fn decode<I: parity_scale_codec::Input>(
+            input: &mut I,
+        ) -> Result<Self, parity_scale_codec::Error> {
+            let NumericScaleHelper { mantissa, scale } = NumericScaleHelper::decode(input)?;
+            match Numeric::try_new(mantissa, scale) {
+                Ok(numeric) => Ok(numeric),
+                Err(NumericError::MantissaTooLarge) => {
+                    Err("error decoding numeric: mantissa too large".into())
+                }
+                Err(NumericError::ScaleTooLarge) => {
+                    Err("error decoding numeric: scale too large".into())
+                }
+                // Not possible to get malformed or negative value from mantissa and scale
+                Err(NumericError::Malformed | NumericError::Negative) => unreachable!(),
+            }
+        }
+    }
+}
+
+mod schema_ {
+    use iroha_schema::{
+        Declaration, Ident, IntoSchema, MetaMap, Metadata, NamedFieldsMeta, TypeId,
+    };
+
+    use super::*;
+
+    impl TypeId for Numeric {
+        fn id() -> Ident {
+            "Numeric".to_string()
+        }
+    }
+
+    impl IntoSchema for Numeric {
+        fn type_name() -> Ident {
+            "Numeric".to_string()
+        }
+
+        fn update_schema_map(metamap: &mut MetaMap) {
+            if !metamap.contains_key::<Self>() {
+                if !metamap.contains_key::<u128>() {
+                    <u128 as iroha_schema::IntoSchema>::update_schema_map(metamap);
+                }
+                if !metamap.contains_key::<u32>() {
+                    <u32 as iroha_schema::IntoSchema>::update_schema_map(metamap);
+                }
+
+                metamap.insert::<Self>(Metadata::Struct(NamedFieldsMeta {
+                    declarations: vec![
+                        Declaration {
+                            name: "mantissa".to_string(),
+                            ty: core::any::TypeId::of::<u128>(),
+                        },
+                        Declaration {
+                            name: "scale".to_string(),
+                            ty: core::any::TypeId::of::<u32>(),
+                        },
+                    ],
+                }));
+            }
+        }
+    }
+}
+
+#[cfg(any(feature = "ffi_export", feature = "ffi_import"))]
+mod ffi {
+    //! Manual implementations of FFI related functionality
+    #![allow(unsafe_code)]
+
+    use super::*;
+
+    iroha_ffi::ffi_type! {
+        unsafe impl Transparent for Numeric {
+            type Target = u128;
```

Comment:
```suggestion
            type Target = [u32; 4];
```

should be fine

Reply:
TLDR: using `[u32; 4]` should be fine.

> and without #[repr(c)] rust doesn't give any guarantees regarding the layout

`#[rerp(transperent)]` is used for `Numeric` so it's layout is exactly the same as for `rust_decimal::Decimal`.

For `Deicmal` feature `repr-c` is enabled by default currently (with @mversic suggestion it would be used for `ffi-export/import` features).

So we know layout of this type, it's size is equal to the size of `u128`.

Neither `u128` nor `Decimal` has any trap representation, so any bit pattern of size 128 would be valid representation of this type. 

So calling `transmute` for this types is fine.

Unfortunately despite [documentation](https://docs.rs/rust_decimal/1.34.2/rust_decimal/#c-repr) authors of the library didn't ensured that alignment is indeed equal to alignment of `u128` :(

This mean that casting pointers would lead to `UB` unfortunately.

Miri shows this as well: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0da091ec2be40ec4866914420749426a).

So i guess using `[u32; 4]` is fine because it's aligment and size are equal to `Decimal` and didn't have any trap representation.

Not extensive proof but Miri is ok: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a2d5974b4db4205215a885415e218275)

In meantime i can open PR into `rust_decimal` with alignment fix.

Reply:
after some thinking I think that, since their struct is `repr(C)`, it would be best to implement:
```rs
impl ReprC for rust_decimal::Decimal {}

impl Transparent for Numeric {
    type Target = rust_decimal::Decimal;
}
```

When a crate provides a C repr for their struct, that's exactly what we should use and not anything else. They are responsible for representation of their struct. However, the orphan rule prevents us from implementing `ReprC` on `rust_decimal::Decimal`. The following [article](https://serde.rs/remote-derive.html) could be of help